ui/boxes: Improve stream sorting for stream typeahead in compose box.

srdeotarse opened the pull request:
<!-- Please see https://github.com/zulip/zulip-terminal#contributor-guidelines ! -->

**What does this PR do?**  <!-- Overall description goes here -->
This PR improves stream sorting for stream typeahead in compose box stream results in footer in the following order -

1) current stream (if matching)
2) pinned and unmuted
3) unpinned and unmuted
4) pinned and muted
5) unpinned and muted

<!-- If fixing a filed bug or new feature, add 'Fixes #<issue>' or 'Partial fix for #<issue>' -->

<!-- Add a link to a discussion on chat.zulip.org, if relevant -->
CZO - https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/Improve.20stream.20sorting.20for.20stream.20typeahead.20in.20compose.20box.2E

**Tested?** <!-- Fine to leave some of these unchecked if this is a draft/work-in-progress -->
- [x] Manually
- [ ] Existing tests (adapted, if necessary)
- [ ] New tests added (for any new behavior)
- [ ] Passed linting & tests (each commit)
<!-- Code must pass CI (GitHub Actions) before merging - look for the green tick! -->
Passing all lint tests.

``test_generic_autocomplete_streams`` test to be updated in ``test_boxes.py``
<!-- See https://github.com/zulip/zulip-terminal#commit-style -->
**Commit flow** <!-- if more than one commit; add/delete/fill-in as appropriate -->
<!-- For example:
- first commit doing some thing
- maybe multiple commits doing similar things
-->

**Notes & Questions** <!-- if any; add/delete/fill-in as appropriate -->
<!-- For example:
- this doesn't include feature X (yet?)
- unsure about Y
- should this do Z?
-->

**Interactions** <!-- if any; add/delete/fill-in as appropriate -->
<!-- eg.
- Waiting on #<PR>
- Blocks #<PR>
-->

**Visual changes** <!-- if any; add/delete/fill-in with screenshot/diagram as appropriate -->
Stream sorting for stream typeahead before -
![image](https://user-images.githubusercontent.com/53873549/162564960-639f268c-b160-426d-8ab6-e4309806a59a.png)

Stream sorting for stream typeahead after -
![image](https://user-images.githubusercontent.com/53873549/162964508-31b452df-8ffd-4d3d-b4e7-a847d8f92f6c.png)

neiljp commented:
Just to note here that as per discussion in the stream, the intention is to leave the search order unaltered during filtering in the left panel (as in the web app zulip/zulip#20618) and adjust the order for other cases such as the stream-box and stream/topic mentions in the main compose box.

srdeotarse commented:
@neiljp as per your note, I have changed the PR topic and updated the code to improve stream sorting in compose box stream typeahead. Recent commit only handles stream typeahead sorting in compose box message area.

srdeotarse commented:
Reopening the PR

srdeotarse commented:
@zulipbot add "PR needs review'.

srdeotarse commented:
@zulipbot add "PR needs review".

neiljp commented on tests/ui_tools/test_boxes.py:
```diff
@@ -819,12 +837,23 @@ def test_generic_autocomplete_streams(
         write_box.view.pinned_streams = streams_to_pin
+        write_box.model.muted_streams = to_mute
+        write_box.stream_id = 1
+        write_box.model.stream_dict = {
```
There is a stream_dict fixture that should be suitable for this.

neiljp commented on tests/ui_tools/test_boxes.py:
```diff
@@ -507,41 +507,23 @@ def test_generic_autocomplete_no_prefix(
+            #     None,
+            #     ["Stream 1", "Stream 2", "Secret stream", "Some general stream"],
+            # ),
+            # ("#NoMatch", None, []),
```
You didn't mention test_generic_autocomplete_set_footer in the stream either, but this is (partially) commented?

Did you miss adding the web public stream?

srdeotarse replied:
I have added web public stream as well in latest commit.

neiljp replied:
I was referring to this commented out code, not the other test cases, which I think you added and were referring to just now?

neiljp commented on tests/ui_tools/test_boxes.py:
```diff
@@ -819,12 +837,23 @@ def test_generic_autocomplete_streams(
         text: str,
         state: Optional[int],
         required_typeahead: Optional[str],
+        stream_id: int,
```
I'm maybe missing something, but you're not using this parameter?

neiljp reviewed (commented):
@srdeotarse Thanks for the update :+1:

The code looks broadly correct, but could be clearer.

I'm a little confused about some of the tests, particularly since the commented out ones appear to be not the ones you named in the stream?

neiljp commented on tests/ui_tools/test_boxes.py:
```diff
@@ -743,88 +725,360 @@ def test_generic_autocomplete_user_mentions(
             (
                 "#Stream",
-                0,
+                1,
```
Why are these existing values changing, with no mutes?

neiljp commented on tests/ui_tools/test_boxes.py:
```diff
@@ -743,88 +725,360 @@ def test_generic_autocomplete_user_mentions(
 
     @pytest.mark.parametrize(
-        "text, state, required_typeahead, to_pin",
+        "text, state, required_typeahead, to_pin, to_mute",
```
To avoid this adding lots of empty data and having lines spill-over, as well as improving the self-documenting of the test cases, I'm wondering if this would be clearer as a dict which might be:
```
{}  # no pins or mutes
{"pinned": [...]}  # equivalent to to_pin=[...]
{"muted": {...}}  # equivalent to to_mute={...}
```

If so, this logic could be a quick refactor commit before this with data and logic adjusted for just to_pin there, and then extended in this commit.

neiljp reviewed (commented):
@srdeotarse Here's another pass.

Assuming no streams are muted, the existing test cases should pass, and shouldn't need altering - unless there is something I'm missing? You might find it useful to split the muting test cases into a separate later commit for now, so you can check that the existing ones pass in one commit and work on others in the next (using rebase or checkout to jump between commits). We may combine the commits later, but that's easier than splitting them apart and wondering why lots of commits fail :)

srdeotarse commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -636,10 +636,31 @@ def autocomplete_streams(
+            for stream_name in matched_streams
+            if stream_name in muted_stream_names
+        ]
+        for stream_name in muted_stream_names:
+            matched_streams.remove(stream_name)
+            if stream_name in [stream["name"] for stream in self.view.pinned_streams]:
+                muted_stream_names.remove(stream_name)
+                muted_stream_names.insert(0, stream_name)
+        matched_streams.extend(muted_stream_names)
```
- When I was writing more test cases, I found that ``muted_stream_ids`` are received in ascending order and not the order of ``left_streams_bar`` (pinned_muted_id, unpinned_muted_id). 
- Therefore had to arrange ``muted_stream_names`` with same order as ``matched_streams`` and then remove ``pinned_muted_streams``, add to start of ``muted_stream_names``.
-  At last, extend it to ``matched_streams`` after removing common streams.

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -636,10 +636,34 @@ def autocomplete_streams(
+                muted_stream_names.insert(0, muted_stream)
+        matched_streams.extend(muted_stream_names)
+
+        current_stream_name = self.model.stream_dict[self.stream_id]["name"]
```
This fails for private messages, and actually causes a crash. There is an outstanding question of what should be the stream to use here - narrow? compose? recent? neither?

neiljp reviewed (commented):
@srdeotarse Thanks for giving the test refactor a try - I left some inline code suggestions that may explain what I had in mind better.

We really need to address the no-stream case, which explains some of the test changes I asked about earlier I believe? Let's discuss the best approach in the topic.

neiljp commented on tests/ui_tools/test_boxes.py:
```diff
@@ -743,74 +743,105 @@ def test_generic_autocomplete_user_mentions(
+                0,
+                "#**Stream 1**",
+                {"pinned": ["Secret stream", "Stream 1"]},
+            ),  # 1st-word startswith match (pinned).
```
You seem to have added these comments? Without them the test cases fit on one line still?

neiljp commented on tests/ui_tools/test_boxes.py:
```diff
@@ -743,74 +743,105 @@ def test_generic_autocomplete_user_mentions(
                 "#**Stream 2**",
-                ["Secret stream"],
+                {"pinned": ["Secret stream"]},
             ),  # 1st-word startswith match (unpinned).
```
I think it may be clearer if we remove these comments in a separate commit (maybe when adding the web public cases?).

However, see my general review point too.

neiljp reviewed (commented):
@srdeotarse I've mainly looked at the first part of this for now.

Your added tests may well be good, but they're difficult to review since they're not in order like others (eg. 0,1,2,3,4 vs 0,1,4,2,3 or similar).

While this is a relatively small thing to put a lot of effort into, one thought to simplify the tests further would be to describe sequences of matches for a range of indices, which would reduce the need to repeat the search text and stream properties. I'm wary of putting extra logic into the test body and having multiple test cases in one, but this could be more expressive - one alternative may be using separate list comprehensions for each set of test cases and adding them together.

neiljp reviewed (commented):
@srdeotarse The test refactors clean things up very well - and maybe other similar ones would benefit?

I'd like to merge the early commits, so if you split out the web-public change and address any comments related to those commits, we can get those in first. At least one of the later commits seem to be failing, which may be related to my thought that the output is reversed?

One benefit of the muted/pinned groups are that they're sort of self-documenting, so some of the comments are redundant. That said, the muted streams are by id right now - maybe if we express the tests by name instead then this would be even clearer?

Generally if a single test case has a dedicated comment, then it's worth using an id string instead, since test passes (and failures!) are then much more descriptive.

Rohitth007 commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -636,10 +636,29 @@ def autocomplete_streams(
+        matched_streams.extend(muted_stream_names)
+
+        matched_stream_typeaheads = format_string(matched_streams, "#**{}**")
+        return matched_stream_typeaheads, matched_streams
```
I haven't followed this PR a lot but I feel that all of this code should belong inside `match_stream`. Is there any reason for not doing so? We can send in the model variables as arguments into the function and process everything there.

Rohitth007 replied:
@neiljp On a design level, I wanted to ask why, in the model, we treat `muted_streams` different from `(un)pinned_streams`, i.e., different data types.

srdeotarse replied:
And about data type of ``muted_streams``, I agree with using ``stream_names`` instead of ``stream_ids`` like we have implemented for ``(un)pinned_streams``

Rohitth007 replied:
Oh I see, but a boolean like `prioritize_unmuted` might help? Since `match_stream` also does some sorting apart from matching, it's better to place all sorting code together.
* If False, it'll return pinned matches then unpinned matches.
* If True, it'll return pinned unmuted matches, unpinned unmuted matches, pinned muted matches then unpinned muted matches.

neiljp replied:
@Rohitth007 I agree there are multiple ways of writing the code, and we may even update the stream order later - or add options - but for now I'm happy for this to be focused on the functionality and ensuring this behaves as expected by using the new tests (and test structure). Once we have tests, we can refactor more safely. I think @srdeotarse already rewrote the code, and we still may need to adjust the algorithm.

neiljp replied:
Re muted_streams vs (un)pinned_streams, these are data-structures used for different purposes right now. The former is simply to know if a stream is muted, while the latter is designed for the UI (historically).

Regarding the tests, I would favor having names in the test cases to make them more readable, independently of the data in the model. That is eg. `{"pinned": {"Stream 1"}, "muted": {"Secret Stream"}}` is more readable than the latter being an int.

Rohitth007 replied:
The reason for having it inside `match_stream` is it would simplify the logic by a lot. Here we are throwing away the matched typeaheads, resorting an already sorted list given by `match_streams`  and then recreating new matched typeaheads. If this is done inside `match_stream` then it would barely take 5-6 lines and it wouldn't affect any of the tests.

Another reason is we would barely have to do any work to extend this feature to the stream panel.

srdeotarse commented on tests/ui_tools/test_boxes.py:
```diff
@@ -814,6 +814,66 @@ def test_generic_autocomplete_user_mentions(
             ),
+            # With 'Secret stream' muted.
+            (
+                "#Stream",
```
3 tests started failing in ``test_helper.py`` after I rebased my current branch with ``upstream/main``. Test cases are as follows -
* ``test_download_media``
* ``test_open_media[0-error0]``
* ``test_open_media[1-error1]``

neiljp replied:
I haven't seen these failures locally after rebasing, only the new tests you added here. Do you still get the errors you saw? (Was that locally or CI, ie. GitHub actions?)

To clarify, it seems that it's one test from each group, perhaps, or at least not an entire case that's failing. While the assert-in-a-loop does detect the error, as I mentioned previously it may be more useful to show the failing pairs all together, which might be straightforward if you eg. build a dict and compare it to that expected?

These new tests apply to the new code, so this should go with the previous commit that changed/added the functionality.

neiljp commented on tests/ui_tools/test_boxes.py:
```diff
@@ -743,57 +743,72 @@ def test_generic_autocomplete_user_mentions(
         )
```
In the commit introducing stream_categories, note that while details can be useful, in the commit text it's more useful to write what you can't easily include in the commit (code) itself, such as why the change is useful/important.

Some of that why might be noting how it might be extended, but until the extra parameters are added, it's not important to refer to them by name - otherwise you couple this commit text to how it *will* be implemented, and either this commit text needs updating later (though it might be merged first!), or else the later code 'must' be named appropriately later (which one cannot always guarantee). Basically, it can be useful to describe what else it might be applied to, or perhaps an example, but otherwise focus the commit text on *this* commit.

I may have given an impression that something wasn't in the commit text which led to this; I believe I was referring to how the commit text didn't describe what this commit was doing (re switching to stream_categories).

neiljp commented on tests/ui_tools/test_boxes.py:
```diff
@@ -815,6 +815,66 @@ def test_generic_autocomplete_user_mentions(
+                    4: "#**Stream 2**",
+                },
+                {"pinned": ["Secret stream", "Stream 1"], "muted": {99, 2}},
+            ),
```
In this commit where you add the test cases, you're not connecting the muted elements to the muted_streams data.

neiljp replied:
Yes, it's there now, but it had disappeared in the previous iteration it seemed.

neiljp reviewed (commented):
@srdeotarse I focused on reviewing the first 5 commits, since the first ones are close to ready and hopefully the muting one will be when the test situation is clarified :)

neiljp commented:
@srdeotarse I've just pushed the first 3 commits here, with some changes to the commit text reasoning as those ending in https://github.com/zulip/zulip-terminal/commit/898bb2e3722567f6a4f4bcc0730f2ad77cfa9d9f. Rebasing should clear any conflict.

I've not looked at more at this point, as the corresponding tests for the first code change fail, so there is some mismatch between test expectations and code behavior.

neiljp reviewed (commented):
@srdeotarse We discussed the changes in the stream, so I'm merging shortly. I had some outstanding comments so I'm posting this as a review.

neiljp commented:
@srdeotarse As per our discussion in the stream, I squashed your feature code+test commits together, and added some adjustments as per the extra commits, resulting in the branch with separated commits at https://github.com/neiljp/zulip-terminal/tree/review-original-1186-6-with-adjust

I did a final tidy of the commit text, to emphasize the action associated with the feature commits, ie. what changed/why, which you can see by comparison with your original commits in the link above or that you may have locally. This kind of detail can be helpful not just at this stage for merging and reading later, but also for reviewers wanting to know "What's in this commit?" and whether it matches the description.

Thanks for pushing this to completion - as I say, merging now! 🎉 
(with those fixup commits squashed into yours)
